### PR TITLE
Task-6: BE: Implement login user Post API

### PR DESCRIPTION
**Pull request by denys-rybachok**

### CODE REVIEW

File: src/constants/errorMessages.constant.ts [hunk 1]
```diff
@@ -1,3 +1,5 @@
 export const statusMessages = {
-  USER_ALREADY_EXISTS: 'User with this email already exists'
+  USER_ALREADY_EXISTS: 'User with this email already exists',
+  INCORRECT_PASSWORD: 'Incorrect password',
+  USER_NOT_FOUND: 'User not found'
```
Comment (VYaniv) [inline src/constants/errorMessages.constant.ts]: Not found message can be reusable and dynamic. Even more, all of them can be dynamic.

File: src/entities/auth/controller/auth.controller.ts [hunk 2]
```diff
@@ -16,4 +17,12 @@ export class AuthController {
   signUp(@Body() userData: CreateUserDto): Promise<{access_token: string}> {
     return this.authService.signUp(userData);
   }
+
+  @ApiOperation({summary: 'User sign-in', description: 'Sign in'})
+  @HttpCode(HttpStatus.OK)
+  @Post('signIn')
+  @UsePipes(new YupValidationPipe(signInUserSchema))
+  signIn(@Body() {email, password}: {email: string; password: string}): Promise<{access_token: string}> {
```
Comment (VYaniv) [inline src/entities/auth/controller/auth.controller.ts]: Create a dto for login request.

File: src/entities/auth/service/auth.service.ts [hunk 3]
```diff
@@ -24,4 +24,20 @@ export class AuthService {
...
+  async signIn(email: string, password: string): Promise<{access_token: string}> {
+    const user = await this.userService.findOne(email);
+    if (!user) {
+      throw new HttpException(statusMessages.USER_NOT_FOUND, HttpStatus.NOT_FOUND);
+    }
+    const isValidPass = await this.userService.checkPassword(password, user.password);
+    if (!isValidPass) {
+      throw new HttpException(statusMessages.INCORRECT_PASSWORD, HttpStatus.BAD_REQUEST);
+    }
+    const payload = {sub: user._id};
```
Comment (VYaniv) [inline src/entities/auth/service/auth.service.ts:37]: Why `sub`?

File: src/entities/user/service/user.service.ts [hunk 4]
```diff
@@ -24,4 +25,8 @@ export class UserService {
     const savedUser = await newUser.save();
     return savedUser;
```
Comment (VYaniv) [inline src/entities/user/service/user.service.ts]: `return await newUser.save();`

File: src/entities/user/validation/signInUserValidation.schema.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import {validationErrors} from '@app/constants/validationErrors.constant';
+
+export const signInUserSchema = yup.object().shape({
+  email: yup.string().email(validationErrors.INVALID_VALUE).required(validationErrors.FIELD_REQUIRED),
+  password: yup
+    .string()
+    .matches(/[a-z].*[a-z]/, validationErrors.PASSWORD_INVALID_LOWERCASE_VALUE)
+    .matches(/[A-Z].*[A-Z]/, validationErrors.PASSWORD_INVALID_UPPERCASE_VALUE)
+    .matches(/[0-9].*[0-9]/, validationErrors.PASSWORD_INVALID_NUMBERS_VALUE)
+    .matches(/[!@#$%^&*()_+\\[\]{};':"|,.<>?].*[!@#$%^&*()_+\\[\]{};':"|,.<>?]/, validationErrors.PASSWORD_INVALID_SYMBOLS_VALUE)
```
Comment (VYaniv) [inline src/entities/user/validation/signInUserValidation.schema.ts]: Constant for regular expressions.

**Review by VYaniv — APPROVED** [review]

File: src/entities/auth/service/auth.service.ts (hunk 3 above)
Comment (denys-rybachok) [reply src/entities/auth/service/auth.service.ts:37]: It is according to JWT standards: "In the JSON Web Token (JWT) standard, the "sub" (subject) claim is a string that identifies the principal that is the subject of the JWT.  The "sub" claim is used to provide some basic information about the context in which the JWT was issued."

File: src/constants/errorMessages.constant.ts (hunk 1 above)
Comment (denys-rybachok) [reply src/constants/errorMessages.constant.ts]: Do you mean like this? To clarify which user or id is incorrect. 
```
USER_NOT_FOUND: (email: string) => `User with email ${email} not found`, 
USER_ALREADY_EXISTS: (email: string) => `User with email ${email} already exists`,
```

File: src/entities/auth/controller/auth.controller.ts (hunk 2 above)
Comment (denys-rybachok) [reply src/entities/auth/controller/auth.controller.ts]: Done.
```
export class LoginUserDto {
  @ApiProperty()
  email: string;

  @ApiProperty()
  password: string;
}
```

File: src/constants/errorMessages.constant.ts (hunk 1 above)
Comment (VYaniv) [reply src/constants/errorMessages.constant.ts]: It is also possible solution but I mean `(entity: string, something: string) => `${entity} with ${something} already exists` .Something like this.

File: src/entities/auth/service/auth.service.ts (hunk 3 above)
Comment (VYaniv) [reply src/entities/auth/service/auth.service.ts:37]: Okay, let`s skip it for now. Thanks.

**Pull request merged**